File: python/mincut/mincut.py
#!usr/bin/env python
#
# Karger's Min Cut algorithms 
#
import sys
import random

class Vertex:
    def __init__(self):
        self.label = "" 
        self.edges = []

    def __str__(self):
        sout = "Label: {label}\n".format(label=self.label)
        for e in self.edges:
            sout += str(e)
        sout += "\n"
        return sout

class Edge:
    def __init__(self):
        self.v1 = None
        self.v2 = None

    def __init__(self, v1, v2):
        self.v1 = v1
        self.v2 = v2

    def __str__(self):
        sout = " {v1} -> {v2}".format(v1=self.v1, v2=self.v2)
        return sout

class Graph:
    def __init__(self):
        self.V = {} 
        self.E = []

    def contract(self, i):
        edges = []
        se = self.E[i]
        v1 = se.v1
        v2 = se.v2

        # move edges from v2 to v1  
        for j in range(len(self.V[v2].edges)):
            if v1 != self.V[v2].edges[j].v2:
                e = Edge(v1, self.V[v2].edges[j].v2)
                self.V[v1].edges.append(e)

        # update vertices to point to merged vertex
        for k, v in self.V.iteritems():
            edges = []
            for j in range(len(self.V[k].edges)):
                e = None
                if self.V[k].edges[j].v2 == v2:
                    e = Edge(self.V[k].edges[j].v1, v1)
                else:
                    e = Edge(self.V[k].edges[j].v1, self.V[k].edges[j].v2)
                edges.append(e)
            self.V[k].edges = edges

        # remove self loops in vertex edges
        self.V[v1].edges = [ e for e in self.V[v1].edges if e.v1 != e.v2 ] 
       
        # update the edges
        for j in range(len(self.E)):
            if self.E[j].v1 == v2:
                self.E[j].v1 = v1 
            elif self.E[j].v2 == v2:
                self.E[j].v2 = v1
        
        # remove self loops in edges
        self.E = [ e for e in self.E if e.v1 != e.v2 ]

        # remove v2 vertex         
        del self.V[v2]


    def randomized_contraction(self):
        while len(self.V) > 2:
            i = random.randint(0,len(self.E)-1)
            self.contract(i)
        return len(self.E) / 2 
    
    def __str__(self):
        sout = ""
        for k,v in self.V.iteritems():
            sout += str(v)
        sout += "Edges:\n"
        for e in self.E:
            sout += str(e)
        sout += "\n"
        return sout

def create_graph(table):
    g = Graph()
    #   init vertices 
    for row in table:
        v = Vertex()
        v.label = row[0]
        g.V[row[0]] = v

    for i in range(len(table)):
        for j in range(1,len(table[i])):
            v1_label = g.V[table[i][0]].label
            v2_label = g.V[table[i][j]].label
            e = Edge(v1_label, v2_label)
            g.V[table[i][0]].edges.append(e)
            g.E.append(e)
    return g

if __name__ == "__main__":
    g = Graph()

File: go/src/mincut/mincut.go
/*
	mincuts
	Kargar's min cut algorithm
*/
package mincut

import (
	"fmt"
	"math/rand"
)

type Edge struct {
	v1 string
	v2 string
}

type Vertex struct {
	label string
	edges []*Edge
}

type Graph struct {
	V map[string]*Vertex
	E []*Edge
}

func NewGraph(t [][]string) *Graph {
	g := new(Graph)

	// initialize vertices
	g.V = make(map[string]*Vertex)
	for i := 0; i < len(t); i++ {
		if t[i][0] == "" {
			panic(fmt.Sprintf("Invalid table::[%d][%d]\n", i, 0))
		}
		g.V[t[i][0]] = new(Vertex)
		g.V[t[i][0]].label = string(t[i][0])
		g.V[t[i][0]].edges = make([]*Edge, 0)
	}

	// initialize edges
	g.E = make([]*Edge, 0)
	for i := 0; i < len(t); i++ {
		for j := 1; j < len(t[i]); j++ {
			e := new(Edge)
			e.v1 = t[i][0]
			if t[i][j] == "" {
				panic(fmt.Sprintf("Invalid table::[%d][%d]\n", i, j))
			}
			e.v2 = t[i][j]
			g.E = append(g.E, e)
			g.V[t[i][0]].edges = append(g.V[t[i][0]].edges, e)
		}
	}
	return g
}

func (g *Graph) RandomizedContraction() int {
	i := 0
	for len(g.V) > 2 {
		i = rand.Intn(len(g.E))
		g.Contract(i)
	}

	return int(len(g.E) / 2)
}

func (g *Graph) Contract(index int) int {

	if index >= len(g.E) {
		return 1
	}
	se := g.E[index]
	v1 := se.v1
	v2 := se.v2

	result := g.transferVertexEdges(v1, v2)
	if result == 0 {
		g.adjustVertexEdges(v1, v2)
		g.removeSelfLoops(v1)
		g.updateEdges(v1, v2)
		g.removeVertex(v2)
	}
	return 0
}

func (g *Graph) transferVertexEdges(v1 string, v2 string) int {
	vertex1 := g.getVertex(v1)
	vertex2 := g.getVertex(v2)
	if vertex1 == nil || vertex2 == nil {
		return 1
	}
	// move edges from v2 to v1
	edges := make([]*Edge, 0)
	for i := 0; i < len(g.V[v2].edges); i++ {
		if g.V[v2].edges[i].v2 == v1 {
			continue
		}
		e := new(Edge)
		e.v1 = v1
		e.v2 = g.V[v2].edges[i].v2
		edges = append(edges, e)
	}
	g.V[v1].edges = append(g.V[v1].edges, edges...)
	return 0
}

func (g *Graph) adjustVertexEdges(v1 string, v2 string) {
	// adjust v2 on the edges on each vertex
	for k, _ := range g.V {
		edges := make([]*Edge, 0)
		for i := 0; i < len(g.V[k].edges); i++ {
			e := new(Edge)
			if g.V[k].edges[i].v2 == v2 {
				e.v1 = g.V[k].edges[i].v1
				e.v2 = v1
			} else {
				e.v1 = g.V[k].edges[i].v1
				e.v2 = g.V[k].edges[i].v2
			}
			edges = append(edges, e)
		}
		g.V[k].edges = edges
	}
}

func (g *Graph) removeSelfLoops(v1 string) {
	// remove self loops
	edges := make([]*Edge, 0)
	for i := 0; i < len(g.V[v1].edges); i++ {
		if g.V[v1].edges[i].v2 == v1 {
			continue
		}
		e := new(Edge)
		e.v1 = g.V[v1].edges[i].v1
		e.v2 = g.V[v1].edges[i].v2
		edges = append(edges, e)
	}
	g.V[v1].edges = edges
}

func (g *Graph) updateEdges(v1 string, v2 string) {
	// update graph edges array
	edges1 := make([]*Edge, 0)
	for i := 0; i < len(g.E); i++ {
		e := new(Edge)
		if g.E[i].v1 == v2 {
			e.v1 = v1
			e.v2 = g.E[i].v2
		} else if g.E[i].v2 == v2 {
			e.v1 = g.E[i].v1
			e.v2 = v1
		} else {
			e.v1 = g.E[i].v1
			e.v2 = g.E[i].v2
		}
		edges1 = append(edges1, e)
	}
	g.E = edges1

	// remove self loops in edges
	edges2 := make([]*Edge, 0)
	for _, e := range g.E {
		if e.v1 != e.v2 {
			ne := new(Edge)
			ne.v1 = e.v1
			ne.v2 = e.v2
			edges2 = append(edges2, ne)
		}
	}
	g.E = edges2

}

func (g *Graph) removeVertex(v string) {
	// remove vertex
	delete(g.V, v)
}

func (g *Graph) getVertex(key string) *Vertex {
	v, ok := g.V[key]
	if !ok {
		return nil
	}
	return v
}

func (g *Graph) getEdge(i int) *Edge {
	return g.E[i]
}

func (g *Graph) Print() {

	fmt.Println("********************************************************************************")
	for _, value := range g.V {
		value.Print()
	}
	fmt.Println("All edges:")
	for _, e := range g.E {
		e.Print()
	}
	fmt.Println()
	fmt.Println("********************************************************************************")
}

func (v *Vertex) Print() {
	fmt.Printf("Vertex: %s\n", v.label)
	for _, e := range v.edges {
		e.Print()
	}
	fmt.Println()
}

func (e *Edge) Print() {
	fmt.Printf(" {%s, %s}", e.v1, e.v2)
}
